### PR TITLE
Add methods for dealing with fairness in rankings

### LINKED ISSUE #455: Add methods for dealing with fairness in rankings
There has been a fair amount of recent interest in fairness in rankings, e.g., in HR automation. Companies such as LinkedIn or Workable.com produce short-lists of candidates for recruiters and suggest suitable openings to candidates. The fairness thereof has been discussed e.g. by:
https://arxiv.org/abs/1905.01989 (KDD 2019)
and in the recent survey:
https://link.springer.com/article/10.1007/s00778-021-00697-y (The VLDB Journal, 2022).
While some of the support for rankings is already in AI Fairness 360, e.g., via the "ordinal" mode in bias detectors, it would be great to see further support for fairness in rankings in terms of post-processing etc.

### PR DESCRIPTION
**Pull request by andrewklayk**
We've implemented a set of post-processing algorithms for dealing with fairness in rankings, as described in https://arxiv.org/abs/1905.01989 (KDD 2019), complete with a demo notebook and tests.
Along with the algorithms, we've added two useful metrics for evaluating rankings that were introduced or referenced in the paper (Infeasibility Index and Discounted Cumulative Gain). Further details are available in the demo notebook.

closes #455

### CODE REVIEW

File: aif360/algorithms/postprocessing/deterministic_reranking.py
Comment (hoffmansc) [inline aif360/algorithms/postprocessing/deterministic_reranking.py:1]: Can you add a line to [`docs/source/modules/algorithms.rst`](https://github.com/Trusted-AI/AIF360/blob/master/docs/source/modules/algorithms.rst) in order to show the API documentation for this?

File: examples/demo_deterministic_reranking.ipynb [hunk 1]
```diff
@@ -0,0 +1,1103 @@
...
+    "# keep the un-normalized scores for clarity; RegressionDataset normalizes them to be from 0 to 1.\n",
+    "balls_ds.labels = np.transpose([balls['score']])"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To initialize the DeterministicReranking class, we need to pass the **protected attribute values** of the priviliged and unprivileged groups.\n",
```
Comment (hoffmansc) [inline examples/demo_deterministic_reranking.ipynb:224]: "privileged" is misspelled a couple places

**Review by hoffmansc — CHANGES_REQUESTED** [review]
thanks @andrewklayk, this looks great! Other than the small comments, I think we can merge this.

I would love to include this in the `aif360.sklearn` subpackage as well. There is a slightly different API for those algorithms/metrics but it should actually be easier to use since you rely on DataFrames anyway. That can be a separate PR if you prefer or you can add it to this one.

File: aif360/algorithms/postprocessing/deterministic_reranking.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    """A collection of algorithms for construction of fair ranked candidate lists. [1]_ .
+
+    References:
+        .. [1] KDD '19: Proceedings of the 25th ACM SIGKDD 
+           International Conference on Knowledge Discovery & Data Mining, July 2019, Pages 2221-2231.
+    """
+
+    def __init__(self,
+            unprivileged_groups: list[dict],
+            privileged_groups: list[dict]):
```
Comment (hoffmansc) [inline aif360/algorithms/postprocessing/deterministic_reranking.py:17]: This type hint doesn't work in Python 3.8. You can change it to use `List` from `typing`. Type hints are optional for this package, though. We just require the type in the docstring

File: examples/demo_deterministic_reranking.ipynb [hunk 3]
```diff
@@ -0,0 +1,1103 @@
...
+    "    response = requests.get(\"http://www.seaphe.org/databases/LSAC/LSAC_SAS.zip\")\n",
+    "    temp_file_name = os.path.join(temp_dir, \"LSAC_SAS.zip\")\n",
+    "    with open(temp_file_name, \"wb\") as temp_file:\n",
+    "        temp_file.write(response.content)\n",
+    "    with zipfile.ZipFile(temp_file_name, 'r') as zip_ref:\n",
+    "        zip_ref.extractall(temp_dir)\n",
+    "    data = pd.read_sas(os.path.join(temp_dir, \"lsac.sas7bdat\"))\n",
+    "    data = data.assign(gender=(data[\"gender\"] == b\"male\") * 1)\n",
+    "    data['race'] = data['race1']\n",
+    "    data = data[['race', 'gender', 'lsat', 'ugpa', 'zfygpa']]\n"
```
[the remark this answers is not in the document]
Comment (andrewklayk) [reply examples/demo_deterministic_reranking.ipynb:384]: That was the initial idea, but I couldn't get the `tempeh` import to work, despite having installed the `.[LawSchoolGPA]` option of the package. I've tried installing `tempeh` separately, reinstalling AIF360 manually and so on, but to no end. Without the import, neither the `LawSchoolGPADataset` nor the `fetch_lawschool_gpa` work.

[the remark this answers is not in the document]
Comment (andrewklayk) [reply examples/demo_deterministic_reranking.ipynb:384]: An update on this: it seems like the SEAPHE website is no more (domain expired), which means that neither this code nor the AIF360 functions work.

[the remark this answers is not in the document]
Comment (hoffmansc) [reply examples/demo_deterministic_reranking.ipynb:384]: that is unfortunate... I've reached out to SEAPHE and they're investigating. Let's give it a week and see if they can get the website back up.

**Comment by andrewklayk** [comment]
Hello @hoffmansc
I've addressed the issues you stated, except for the usage of the SEAPHE dataset. The site is still down, and I've been unable to find a free-use dataset from a reputable source that would fit here as an example. 
In my opinion, the "toy example" in the notebook is enough to explain the API, so perhaps we should ditch the dataset part alltogether?

**Comment by rahulnair23** [comment]
Hi @hoffmansc - seems like the Law School dependency has been removed from the demo notebook. Is this good to merge?

**Review by hoffmansc — APPROVED** [review]

**Pull request merged**